Styles: try wrapping with :root to fix reset styles

ellatrix opened the pull request:
<!-- Thanks for contributing to Gutenberg! Please follow the Gutenberg Contributing Guidelines:
https://github.com/WordPress/gutenberg/blob/trunk/CONTRIBUTING.md -->

## What?
<!-- In a few words, what is the PR actually doing? -->

In order to apply global styles _after_ any theme reset styles, e.g. `ol, ul { ... }`, increase the specificity _slightly_, while _keeping_ CSS overall specificity levelling and specificity decreases.

To do: figure out where else exactly we need `:root`, I'm not super familiar with this code.

_**Note: Block style variations and their selectors will be handled via a separate follow-up as they haven't yet had their specificity reduction merged in Gutenberg (see https://github.com/WordPress/gutenberg/pull/61032).**_

## Why?
<!-- Why is this PR necessary? What problem is it solving? Reference any existing previous issue(s) or PR(s), but please add a short summary here, too -->

To make sure reset styles not wrapped in `:where` do not break anything.

Why `:root` and not `html` or `body`? Because `:root` has the same specificity of attributes and would catch all kinds of reset styles including those with attributes ([example](https://github.com/necolas/normalize.css/blob/fc091cce1534909334c1911709a39c22d406977b/normalize.css)).

## How?
<!-- How is your PR addressing the issue at hand? What are the implementation details? -->

Wrapping all our specificity normalised/levelled rules in `:root`, so that they run after these reset styles.

## Testing Instructions

The only unit tests we have are covering the generation of styles rather than their application, so we'll have to rely on some manual testing or adding new tests. Any help here is appreciated! :)

#### Manual Testing

- Activate emptytheme
- Add all blocks to a post or page in both editors, ensure they display correctly and block library styles are applied
- Switch to the switch editor and apply global styles for each block, check these all take affect both in the editor and frontend
- Test various classic themes making sure the blocks get all the correct styles on the frontend or in the editors


### Testing Instructions for Keyboard
<!-- How can you test the changes by using the keyboard only? Please note, this is required for PRs that change the user interface (UI). This ensures the PR can be tested for any possible accessibility regressions. -->

## Screenshots or screencast <!-- if applicable -->

ellatrix committed:
Styles: try wrapping with :root to fix reset styles

ellatrix committed:
Adjust based on #59457

tellthemachines reviewed (commented):
Thanks for the PR! 

In terms of specificity, `:root` is exactly the same as a classname, so this change effectively undoes the reduction in #60106. I think that if that's what we want to do, reverting #60106 is a better solution as it will keep the simpler, more semantic classname selector.

The other thing to bear in mind is that, if we revert the global styles specificity, we'll need to revert #60228 too, because the reduction in the layout selectors was dependent on the global styles one.

I don't have enough context on why this change is needed, so a fuller description on this PR, or link to an issue, would be much appreciated!

ellatrix commented:
We don't want to undo any reduction in specificity, we just want to put it at a level that is the same or higher as the reset styles. We don't want to revert global styles specificity reduction.

> I don't have enough context on why this change is needed, so a fuller description on this PR, or link to an issue, would be much appreciated!

I'm not sure how much fuller I can make it, but I'll try. When there are reset styles at the element specify level, they will overrule any zero specificity global styles?

aaronrobertshaw commented:
My understanding of the proposal is that:

- All generated styles still need to have the same specificity
- A revert of [#60106](https://github.com/WordPress/gutenberg/pull/60106) would result in varied selector specificity (see [button block](https://github.com/WordPress/gutenberg/blob/trunk/packages/block-library/src/button/block.json?plain=1#L124))
- This PR would only bump the uniform specificity to a degree that prevents reset styles from overriding block styles

So if that's correct, I don't think a revert is required but there might be some additional tweaks required for the block library styles for core blocks.

This does still provide a general reduction in specificity, unfortunately, not to zero as we would in an ideal world starting from scratch.

ellatrix committed:
Fix php unit tests

ellatrix committed:
Fix js tests

ellatrix committed:
Add a basic test for overriding reset styles

ellatrix committed:
Add test for global styles override

ellatrix committed:
Don't add :root :where for single classes, because they are already 0-1-0 specificity

ellatrix commented on packages/block-library/src/template-part/theme.scss:
```diff
@@ -1,9 +1,7 @@
 // Same as the group block styles.
-:where(.wp-block-template-part) {
-	&.has-background {
```
It seems we previously did have some specificity left here. We should make sure it tests well now.

ellatrix commented on packages/block-library/src/gallery/editor.scss:
```diff
@@ -1,4 +1,7 @@
+// Adding `figure` to the selector increases the specificity above the global
+// styles specificity, which is levelled at 0-1-0. We should figure out why
+// `figure` is needed.
+:root :where(figure.wp-block-gallery) {
```
I just added `:root` as a precaution here to make sure this overrides any reset styles.

aaronrobertshaw commented:
There were some additional styles reduced to zero specificity in https://github.com/WordPress/gutenberg/pull/47858. It appears they might also need prefixing with `:root`.

aaronrobertshaw committed:
Try adding :root to layout stykes

aaronrobertshaw commented:
I've pushed a commit that I think might be the minimum set of changes required to make layout styles work and maintain some more consistent specificity.

To be honest, I had a really hard time working out what should and should not be updated for the layout style selectors. Especially, when running into bugs that turned out to also be on trunk (#61849, #61846 etc)

Moving forward I think we definitely need some test coverage to ensure the correct layout styles are being applied when and where they should. These would also serve as internal documentation for anyone needing to change some aspect of our layout supports.

It's also worth noting here that this PR is being discussed in regards to what [CSS specificity we should settle on for WP 6.6](https://github.com/WordPress/gutenberg/discussions/61810).

aaronrobertshaw commented:
After some [discussion](https://github.com/WordPress/gutenberg/discussions/61810), and taking into account the limited time before the 6.6 beta, the option to use a `:root` prefix seems the safest bet and appears to have majority support.

This PR should be ready to be refined, reviewed, etc with a view to landing as soon as possible.

talldan commented on packages/block-library/src/embed/theme.scss:
```diff
@@ -2,6 +2,6 @@
 }
 
-:where(.wp-block-embed) {
+.wp-block-embed {
```
Should the `.wp-block-embed figcaption` rule at the top of this file have lowered specificity?

Rather than adding `:root` to each individual rule or removing `:where`, I thought it might be better to do something like this:
```scss
:root {
    :where(.wp-block-embed) {
        // ...
    }
}
```

This way everything is consistent.

I realize it's the same specificity as just a classname, but it makes it easier for anyone modifying the css to maintain the same specificity.

In the future if we ever need to implement layers, we can replace the `:root` with `@layer name`, or if we want to reduce specificity to zero we remove the `:root {}` wrapper.

aaronrobertshaw replied:
> Should the .wp-block-embed figcaption rule at the top of this file have lowered specificity?

I'd say so. A rough rule could be "anything that is configurable via global styles should be".

At a glance that mixin touches `color` and `font-size` rules and captions are now a configurable element in Global Styles.


> Rather than adding :root to each individual rule or removing :where, I thought it might be better to do something like this:

Sounds good to me. I'm still working through the impacts this PR will have on https://github.com/WordPress/gutenberg/pull/61032 and haven't gone through all the block styles yet.

My understanding is this PR has only tweaked what was changed in the original zero-specificity PRs which has proven to be somewhat incomplete. There was some discussion that we needed a complete audit of all the block styles once everything settled.

aaronrobertshaw replied:
One possible issue with the suggested approach to nesting `:root` and `:where` like that is it might encourage people to think they can still just nest the next selector within the `:where` when they really have to define the entire selector or make the nested selector chain another `:where()` e.g.

```scss
:root {
    :where(.wp-block-button) {
        // ...
        &:where(> .wp-block-button__link) {
            // ...
        }
    }
    :where(.wp-block-button > .wp-block-button__link) {
        // ...
    }
}
```

talldan replied:
Yeah, it gets a bit much doesn't it. 😬

talldan replied:
> :root {
>    :where(.wp-block-embed) {
>        // ...
>    }
> }

I had a look at this suggestion, but it becomes convoluted with some of the more complex selectors, so probably best to keep it simple right now.

aaronrobertshaw committed:
Reduce block style variation selector specificity to 0-1-0

aaronrobertshaw committed:
Update block style variation selectors in global styles hook

aaronrobertshaw committed:
Try applying :root :where selectors for custom css

ellatrix committed:
Add backport log entry

aaronrobertshaw committed:
Update button styles

aaronrobertshaw commented:
I've updated the button block styles in https://github.com/WordPress/gutenberg/pull/61638/commits/5a6a6f68be43a4d750cafa552f8ae769c05dd532. It's worth noting that the styles relating to fallback border display have been removed.

Since https://github.com/WordPress/gutenberg/pull/61556 landed, we have access to the merged global styles data in the editor settings. A follow-up is planned to update the border panel in the editor to leverage this data and apply fallbacks only when required. 

An alternative, looking to move the fallback border styles into the `lib/theme.json` custom css for the button block proved unsuccessful. These styles would need to sit between the core's theme.json styles and any by the theme or user. Addressing the border UX through the UI is a better solution and also means we can ship less CSS for each block that has these fallback styles.

aaronrobertshaw committed:
Update image styles

aaronrobertshaw committed:
Update pullquote styles

aaronrobertshaw committed:
Update separator styles

aaronrobertshaw commented:
It appears the Separator block's global styles aren't working with themes that define more than just a single background color for the block in theme.json, including TT4. This is the current behaviour on trunk so I'm not 100% sure if there has been a regression or it was intended this way for some reason. Improving this can be added to the list of follow-ups.

To progress this PR the Separator block's styles were updated in https://github.com/WordPress/gutenberg/pull/61638/commits/2f70fddf455a6dc96c804e5fcf906db98f0bc8f4, just note the odd behaviour when testing.

aaronrobertshaw committed:
Fix separator dots style

aaronrobertshaw committed:
Update site logo styles

aaronrobertshaw committed:
Update social link styles

aaronrobertshaw committed:
Add comment explaining table selector

aaronrobertshaw committed:
Update tag cloud styles

aaronrobertshaw committed:
Move block style variation styles to after block styles

Given they will have level specificity the variation styles should override the global block styles.

aaronrobertshaw committed:
Fix Cover styles blocking global styles

aaronrobertshaw committed:
Fix image caption styles

aaronrobertshaw committed:
Fix latest posts padding styles

aaronrobertshaw committed:
Fix list block's has-background padding

aaronrobertshaw committed:
Fix paragraph block's has-padding style

aaronrobertshaw committed:
Fix site title link colors

aaronrobertshaw commented:
While https://github.com/WordPress/gutenberg/pull/61932 has been in the works I've tried to give most of our blocks a smoke test comparing this PR against 6.5 and trunk (zero-specificity). For the most post things tested consistently between them all.

<details>
<summary>Blocks Tests</summary>

- Archives
- Audio
- Avatar
- Button
- Buttons
- Calendar
- Categories List
- Code
- Column/s
- Cover
- Details
- Embed
- File
- Gallery
- Group
- Heading
- Image
- Latest Comments
- Latest Posts
- List
- List Item
- Media Text
- Navigation
- Navigation Link
- Page List
- Paragraph
- Post Author
- Post Author Biography
- Post Author Name
- Post Date
- Post Featured Image
- Post Terms
- Post Time To Read
- Pullquote
- Quote
- Read More
- RSS
- Search
- Separator
- Site Logo
- Site Tagline
- Social Link/s
- Spacer
- Table
- Tag Cloud
- Term Description
- Verse
- Video
</details>

There were quite a few issues encountered that were pre-existing and present on WP6.5, trunk (zero-specificity) and this PR. 

Some of those were able to be fixed via minor style tweaks on this PR. Nice to be able to improve the status quo a little 🙂 

<details>
<summary>Common / Pre-existing Issues</summary>

- **Archive**: 
   - Block instance styles do not override global styles as they are applied to the wrapper and not the inner ul list
- **Avatar**: 
   - Making the avatar block link to the user profile breaks the layout a bit in the editor. It appears to lose the left/right margins. Block renders correctly on the frontend.
   - The `.wp-block-avatar__image img` style doesn't seem to target anything as the `img` element in the block actually gets the `.wp-block-avatar__image` class. Is this related to an old markup or something?
- **Button**
   - Most block instance typography styles do not override global typography styles for the Button block
- **Buttons**
    - Global font size for Buttons block type doesn't apply over button element styles. Block instance styles do because of a `.has-custom-font-size` class that is applied to the wrapper. Might need updating such that the typography styles use a custom selector to target the inner blocks.
-  **Calendar**
   - Block instance background and text colors don't override global styles or apply generally in the editor.
- **Cover**
   - Regression (introduced with zero specificity). Global element styles for a block type aren't being applied to a heading element (**fixed** https://github.com/WordPress/gutenberg/pull/61638/commits/73539526444827205f0770dd8de8a663af18b4be)
- **Embed**
   - Global margin styles are applying to the inner div but only in the editor.
- **Image**
   - Global caption font size and color styles don't get applied when opinionated theme styles are active (**fixed** https://github.com/WordPress/gutenberg/pull/61638/commits/2ce6eb265ecc1317ed9a363f01b368357f6c13cb)
   - Duotone filters added to rounded block style variation are not applied on the frontend
- **Latest Posts**
   - Block library styles are blocking global padding styles from applying (**fixed** https://github.com/WordPress/gutenberg/pull/61638/commits/1840d3b447b59c01a763942d8fffd4e7f394fa5b)
- **List**
   - Has background class padding overrides global styles padding (**fixed** https://github.com/WordPress/gutenberg/pull/61638/commits/7810a30d817dec27594aa243aecdc0eb2e7f7d97)
   - Checked block style variation isn't registered in block.json so doesn't appear within global styles for configuration
- **Post Featured Image**
   - When padding is applied to the featured image the overlay is out of sync and covers the padding area as well as the image.
- **Post Terms**
   - The Pill block styles aren't configurable via global styles as they aren't listed in block.json
- **Search**
   - There's a large disconnect between the application of global styles and block instance styles but it is consistent on trunk and WP6.5 as well
- **Separator**
   - The separator block style variations are confusing to work out. The global styles offer a background control but this will only be applied for the dots and wide style variations if only a background global style is set nothing else. As TT3 sets more styles for the separator in the theme.json it effectively prevents global styles changing the color of the separator.
- **Site Logo**
   - Rounded style variation's duotone is not applied on the frontend
   - Block instances cannot override the rounder style variations duotone in the editor but the override does work on the frontend
- **Social Links**
   - Block spacing isn't being applied correctly on the frontend. This PR, trunk and 6.5 all work in the editor only.
- **Tag Cloud**
   - Some typography styles that are applied directly to the wrapper do not override global styles.

</details>

The general process followed when testing blocks was to check the following:
1. Block library styles get applied
2. Editor styles override the library styles
3. Block theme styles override library and editor styles
4. Global styles apply correctly
5. Global element styles take effect
6. Block instance styles override global styles
7. Block instance element styles override global element and block styles
8. Block style variations applied correctly
9. Global styles for block style variations override the variation's defaults
10. Block instance styles override the variation and global block style variation styles

As this testing was quite long-winded and tedious, I'm not sure I can guarantee I didn't miss something.

The following blocks and styles could really use some extra eyes:
1. Layout styles
2. Cover block
3. Navigation block

aaronrobertshaw commented:
To get a head start on backporting these changes and all the preceding zero-specificity updates, I've drafted an alternate backport PR (https://github.com/WordPress/wordpress-develop/pull/6633) to cover it all. The backport log in this PR hasn't been updated yet in case we do just roll the alternate backport changes into the original core PR.

aaronrobertshaw commented:
I'm wondering if I might have jumped the gun in https://github.com/WordPress/gutenberg/pull/61638/commits/66cb8f3aa4db57e2340122f164e6c4c526cabee8 & https://github.com/WordPress/gutenberg/pull/61638/commits/362d2ae3fbf3c7d5cb5959bdd26e795243708574.

The changes in those commits allow global styles padding for the blocks to take precedence over the default `.has-background` styles for the list and paragraph blocks.

Imagine the following scenario:

- Theme defines padding styles for the paragraph block within theme.json
- A post has a paragraph block with a background color set (results in `.has-background` class being applied)
- Currently, the theme.json padding wouldn't be applied as the `p.has-background` selector has higher specificity
- If the commits above land, then the padding applied on the paragraph block would change
- If the theme also enqueued its own stylesheet that further customized the `p.has-background` styles, they'd still take effect

[Searching](https://wpdirectory.net/search/01HYVX2ZY76B3ZMV9A2FMHYGPY) WP Directory I only found one theme that touched on `padding` styles for the paragraph block (the example below). So this is definitely an edge case.

It appears some themes have gotten around the `p.has-background` style by defining custom CSS within the paragraph block's theme.json styles. For example by:
```json
	"core/paragraph": {
		"css": ".has-background {padding: 20px 30px;}"
	},
```

Given the spirit of this PR is to reduce risk of broken styles for WP6.6, I'm happy to revert these .has-background tweaks. They could be an easy tweak and follow-up post-6.6. What do you think?

andrewserong reviewed (commented):
Overall this appears to be testing well for me, and thanks for the detailed notes in https://github.com/WordPress/gutenberg/pull/61638#issuecomment-2129191120 @aaronrobertshaw. It seems to me that the overall reduction of global styles is a very complex problem, and I like the pragmatic direction of this PR in that it preserves many of the recent improvements, while adding an additional levelling of styles to find a pragmatic path forward for 6.6. In my testing so far, the layout styles appear to be working well for me, and I haven't encountered any conflicts so far in the flow, constrained, flex, and grid layout types.

Also, some of the changes here appear to resolve some subtle issues for older classic themes that depended on the `wp-block-styles` rules such as the margin reset on the image block.

> I'm wondering if I might have jumped the gun in https://github.com/WordPress/gutenberg/commit/66cb8f3aa4db57e2340122f164e6c4c526cabee8 & https://github.com/WordPress/gutenberg/commit/362d2ae3fbf3c7d5cb5959bdd26e795243708574.

Is the main concern that by making that change, global styles rules might unintentionally override individual paragraph blocks that have set a background color? I'm a little on the fence about this one, as I'd imagine if someone has set the list or paragraph blocks to have a particular value in global styles they might actually want it to override that default `.has-background` padding — my assumption is that that rule was a bit of a pleasing visual reset rather than a desired end state.

> They could be an easy tweak and follow-up post-6.6. What do you think?

That sounds very reasonable to me, and if we find that it is an issue for folks during the beta period, there could be an argument for attempting to get the follow-up in sooner.

aaronrobertshaw committed:
Revert "Fix paragraph block's has-padding style"

This reverts commit 362d2ae3fbf3c7d5cb5959bdd26e795243708574.

aaronrobertshaw committed:
Revert "Fix list block's has-background padding"

This reverts commit 66cb8f3aa4db57e2340122f164e6c4c526cabee8.

aaronrobertshaw commented:
Greatly appreciate the thoughtful review @andrewserong 🙇 

>  I like the pragmatic direction of this PR in that it preserves many of the recent improvements, while adding an additional levelling of styles to find a pragmatic path forward for 6.6

Agreed, the consensus coming from [recent discussions](https://github.com/WordPress/gutenberg/discussions/61810) definitely seems to have settled on a workable approach for WP6.6.

> Also, some of the changes here appear to resolve some subtle issues for older classic themes that depended on the wp-block-styles rules such as the margin reset on the image block.

💯 These subtle issues have proven tricky to surface or predict generally. Another reason why minimising our specificity changes for this release is the safer bet.

> I'd imagine if someone has set the list or paragraph blocks to have a particular value in global styles they might actually want it to override that default .has-background padding — my assumption is that that rule was a bit of a pleasing visual reset rather than a desired end state.

This was my thinking too. I viewed the tweaks as more "fixes" than "changes".

That said, it's a minor edge case that probably doesn't warrant changing now if we don't have to. Giving it a long run in Gutenberg post-6.6 along with some more advanced warning might smooth that transition. 

Another possibility to consider is a user tinkering in Global Styles with the paragraph block padding values. If they left styles there that weren't taking effect due to `.has-background`, those suddenly appearing might be more breaking than carefully crafted values set by the theme author in theme.json.

I've reverted those `.has-background` commits in this PR and will spin up a follow-up for them shortly.

aaronrobertshaw commented:
Separate follow-up for the `.has-background` tweaks is available in https://github.com/WordPress/gutenberg/pull/61999

andrewserong commented on packages/block-library/src/latest-posts/editor.scss:
```diff
@@ -1,8 +1,5 @@
 .wp-block-latest-posts {
 	padding-left: 2.5em;
-	&.is-grid {
-		padding-left: 0;
```
This appears to cause a subtle regression in the editor when using the grid version of the latest posts block, as the above `padding-left: 2.5em` wins out. The added rules in `style.scss` in this PR appear to have lower specificity than that:

| Trunk | This PR |
| --- | --- |
| <img width="1411" alt="image" src="https://github.com/WordPress/gutenberg/assets/14988353/6d803627-55e7-425b-a3e8-2bc716cd86a5"> | ![image](https://github.com/WordPress/gutenberg/assets/14988353/599406ef-bee9-4229-8db1-608dccdab3bb) |

On the site frontend, however, the `padding-left: 0` rule is correct in both trunk and this PR.

andrewserong commented on packages/block-library/src/separator/editor.scss:
```diff
@@ -2,11 +2,4 @@
 	// Prevent margin collapsing so the area to select the separator is bigger.
 	padding-top: 0.1px;
 	padding-bottom: 0.1px;
-
-	// This is also set in style.scss, but needs a higher specificity in editor
-	// due to the way that color block supports adds additional background color styles.
-	&.wp-block-separator.is-style-dots {
-		background: none !important;
-		border: none;
-	}
```
This removal appears to break setting a background color on an individual block in the post editor:

<img width="908" alt="image" src="https://github.com/WordPress/gutenberg/assets/14988353/95926262-cc50-4202-97d4-12b9e5be7972">

It's fine on the site frontend, but not in the editor.

andrewserong reviewed (commented):
Just gave this a re-test, and it's still looking really good to me (tested in block themes and in Twenty Seventeen as a classic theme that uses `wp-block-styles`), with layout styles working where I'd expect them to. Thanks for the detailed follow-ups and considerations here.

I found a couple of small issues in editor styles for a couple of blocks (I've left two comments). Since it only affects the editor, do you think it'd be worth fixing them up in follow-ups? I'm aware this is already a fairly big PR, so I wondered if it'd be easier to address separately (but still for 6.6) since the impact is only within the editor.

Overall the changes here appear to be logical and pragmatic to me, the issues I ran into appear to be to do with the challenges of making changes to fairly brittle styling in some of the core blocks. Typically, I'd lean toward approaching the styling changes for each of these blocks in their own PRs, but given time constraints for the 6.6 feature freeze, in this case, I'm supportive of the idea of going with this PR, and then fixing up issues in subsequent PRs if that works for everyone.

aaronrobertshaw committed:
Fix additional issues with latest posts padding

aaronrobertshaw committed:
Fix enforcement of no borders or background for dots separator

andrewserong reviewed (approved):
Thanks for the updates to the latest posts and separator blocks @aaronrobertshaw, that's resolved those issues for me!

This PR's looking in pretty good shape to me. Given the complexity involved and that in manual testing we believe that things are pretty much working nicely, I think this is in a good place to land, and we can continue to refine things in follow-ups if need be. At this stage, I think merging this larger PR is the pragmatic way to go to ensure the changes can make it in for 6.6, and we still have the Beta period to thorough test the changes and address any points of friction.

LGTM! 🚀

aaronrobertshaw commented:
Appreciate your time and attention to detail on this one @andrewserong 🙇 

I'll look to get this merged then address any follow-ups as required. 

A draft backport PR is available in https://github.com/WordPress/wordpress-develop/pull/6633. It might be easiest to close the currently pending backport for the old zero-specificity changes (https://github.com/WordPress/wordpress-develop/pull/6522) then proceed with the current draft.

aaronrobertshaw commented:
The failed Compressed Size / Check job is only due to the recent stale types issue and unrelated to this PR. I'll go ahead and merge this to save the rebase and re-waiting on e2s, unblocking follow-up work on section styles.

tellthemachines commented:
(belatedly) gave this a test and all seems to be working as expected!

> A draft backport PR is available in [WordPress/wordpress-develop#6633](https://github.com/WordPress/wordpress-develop/pull/6633). It might be easiest to close the currently pending backport for the old zero-specificity changes ([WordPress/wordpress-develop#6522](https://github.com/WordPress/wordpress-develop/pull/6522)) then proceed with the current draft.

That sounds sensible, I went ahead and closed my PR.

aaronrobertshaw commented:
Backport is ready for review over in https://github.com/WordPress/wordpress-develop/pull/6633

aaronrobertshaw commented:
## Dev Note: WordPress 6.6 CSS Specificity

One of the goals of WordPress 6.6 is to simplify the process for theme authors to override core styles while also maintaining support for Global Styles. 

Historically, high CSS specificity in core styles has made customization challenging and unpredictable, often requiring complex CSS rules to achieve desired outcomes. Development of the new [section styles](https://github.com/WordPress/gutenberg/issues/57537) feature also highlighted a need for uniform CSS specificity to support nesting such styles, facilitating the creation of sophisticated, layered designs.

### Uniform 0-1-0 Specificity

WordPress 6.6 introduces several changes aimed at broadly reducing CSS specificity and making it more uniform. These changes generally fall into two categories:

1. Core Block Styles
2. Theme.json / Global Styles:

Where adjustments to CSS specificity were required, they were achieved by wrapping the existing selector within `:root :where(...)`.

#### Core Block Styles

The choice of `0-1-0` specificity greatly reduced the changes required to existing core block styles as blocks targeting their default `.wp-block-<name>` class already have the desired specificity. 

Any blocks with Global Styles support using higher specificity selectors had those selectors wrapped in `:root :where(...)`. This also applied to Block Styles (aka block style variations) and their default styles e.g. `.wp-block-image.is-style-rounded img` was updated to `:root :where(.wp-block-image.is-style-rounded img)`.

#### Theme.json / Global Styles:

All block styles, including block style variation styles, output by Theme.json and Global Styles are now limited to `0-1-0` specificity. Layout styles, e.g. constrained, flex, flow etc., have also been limited however depending on the specific layout type and definition the final specificity varies slightly from `0-1-0` so they apply correctly.

### Usage

The majority of themes and custom blocks should not require any updates.

#### Custom Blocks

Authors of custom blocks that opt into global styles and apply default styling via a selector with greater than `0-1-0` specificity, should update those selectors wrapping them in `:root :where()`.

An example could be a custom list block that opts into padding block support but defines default padding via:

```css
ul.wp-block-custom-list {
    padding: 0;
}
```

Without adjusting the specificity of this rule, any customizations of the block type's padding in Global Styles would be overridden. Wrapping the selector in `:root :where()` here would allow the style load order to determine which rule takes precedence.

```css
// Block's stylesheet
:root :where(ul.wp-block-custom-list) { // This is a contrived example and could simply be `.wp-block-custom-list`
    padding: 0;
}

// Global Styles - Loaded after the block styles
:root :where(.wp-block-custom-list) {
    padding: 1em;
}
```

#### Block Styles (aka Block Style Variations)

Theme authors customizing [Block Styles](https://developer.wordpress.org/block-editor/reference-guides/block-api/block-styles/) for a core block will need to limit their style's specificity so the block style continues to be configurable via Global Styles.

For example, take a theme that tweaks the border radius for the Image block's "rounded" block style:

```css
.wp-block-image.is-style-rounded img {
    border-radius: 2em;
}
```

Without adjustment, this style would override any customizations made to the Rounded block style within Global Styles.

In this case, the theme can tweak its rounded image style to the following:

```css
//. Theme style
:root :where(.wp-block-image.is-style-rounded img) {
    border-radius: 2em;
}

// Global Styles - Loaded after the block styles
:root :where(.wp-block-image.is-style-rounded img) {
    border-radius: 4px;
}
```

### Zero-Specificity, CSS Layers, and The Future

Reducing all core styles to zero specificity was explored before settling on `0-1-0` specificity. Zero specificity unfortunately wasn't robust in the face of common reset stylesheets and required more widespread changes. 

CSS Layers were also evaluated but fell short due to not being able to enforce all styles belonged to a layer. This will change in the future at which point a combination of CSS Layers and zero-specificity can be revisited to further the benefits gained in this initial reduction of CSS specificity.

More history and context can be found in this [detailed discussion](https://github.com/WordPress/gutenberg/discussions/61810).

### Useful Links

- [CSS Specificity for WordPress 6.6](https://github.com/WordPress/gutenberg/discussions/61810)
- [PR CSS reducing specificity](https://github.com/WordPress/gutenberg/pull/61638)
- [Section Styling Feature](https://github.com/WordPress/gutenberg/issues/57537)

justintadlock commented:
> The majority of themes and custom blocks should not require any updates.

I'd be very careful saying this in the Dev Note. So far, I've seen quite a few reports from people building themes where these changes have altered their designs in various ways. I'd say that every theme and block author should check their designs. It doesn't have to be _OMG, the sky is falling_, but it should be proactive in telling them that the reality is that this can and will affect designs where custom CSS comes into play.

aaronrobertshaw commented:
Thanks for the feedback @justintadlock, appreciate it 👍 

While numerous deep dives into the directory showed a majority of themes and plugins shouldn't need an update, that doesn't mean the amount needing tweaks is insignificant. It wasn't my intention to suggest otherwise or create a false sense of security so I've updated that sentence with the following.

> The alignment of 0-1-0 specificity for Global Styles to default block selectors, e.g. `.wp-block-<name>`, greatly reduces the need for updates. It's recommended for theme and block authors to double-check their designs if they rely on custom CSS using more complex selectors.

I hope that strikes a better balance 🤞
